### PR TITLE
cdef: roll loops

### PR DESCRIPTION
**Pull request by tmatth**

### CODE REVIEW

**Comment by tdaede** [comment]
Any difference in performance?

**Comment by tmatth** [comment]
> Any difference in performance?

Not obviously, any tips for benchmarks?

**Comment by tmatth** [comment]
Seems a bit slower (esp. at higher rates):
https://beta.arewecompressedyet.com/?job=master-57112d496c1e95c019a47995df14f07d255bb172&job=cdef-roll-loops%402019-04-24T02%3A12%3A45.066Z

File: src/cdef.rs [hunk 1]
```diff
@@ -151,42 +151,27 @@ unsafe fn cdef_filter_block<T: Pixel>(
...
-        let s3 = *ptr_in.offset(-cdef_directions[(dir + 6) & 7][k]);
-        if s0 != CDEF_VERY_LARGE {
-          max = cmp::max(s0, max);
-        }
-        if s1 != CDEF_VERY_LARGE {
-          max = cmp::max(s1, max);
-        }
-        if s2 != CDEF_VERY_LARGE {
-          max = cmp::max(s2, max);
+        let p = [*ptr_in.offset(cdef_directions[dir][k]),
```
Comment (lu-zero) [inline src/cdef.rs]: while at it `cdef_directions[dir][k]` could be moved out

File: src/cdef.rs [hunk 2]
```diff
@@ -151,42 +151,27 @@ unsafe fn cdef_filter_block<T: Pixel>(
...
+          sum += pri_taps[k] * constrain(i32::cast_from(*p_elem) - i32::cast_from(x), pri_strength, damping);
+          if *p_elem != CDEF_VERY_LARGE {
+            max = cmp::max(*p_elem, max);
+          }
+          min = cmp::min(*p_elem, min);
         }
-        if s3 != CDEF_VERY_LARGE {
-          max = cmp::max(s3, max);
+
+        let s = [*ptr_in.offset(cdef_directions[(dir + 2) & 7][k]),
```
Comment (lu-zero) [inline src/cdef.rs]: Same thing for `cdef_directions[(dir + 2) & 7][k]` and `cdef_directions[(dir + 6) & 7][k]`

File: src/cdef.rs [hunk 3]
```diff
@@ -151,42 +151,27 @@ unsafe fn cdef_filter_block<T: Pixel>(
...
+                 *ptr_in.offset(-cdef_directions[(dir + 2) & 7][k]),
+                 *ptr_in.offset(cdef_directions[(dir + 6) & 7][k]),
+                 *ptr_in.offset(-cdef_directions[(dir + 6) & 7][k])];
+        for s_elem in s.iter() {
+          sum += sec_taps[k] * constrain(i32::cast_from(*s_elem) - i32::cast_from(x), sec_strength, damping);
```
Comment (lu-zero) [inline src/cdef.rs]: `sec_taps[k]` and `pri_taps[k]`  could be out of the loop as well.

**Pull request merged**